### DISCUSSION TITLE
Titles and Subtitles Semantics: Issues and Proposals

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#102] [General]
PMLC 4.0.0

A few consideration on titles and subtitles.

# Doc vs Chapters Subtitles

From the documentation it's not clear whether a PML document supports a subtitle or not, whereas it's clear that chapters do support subtitles.

Most lightweight markup syntaxes only support a subtitle for the document main title, not for chapters/sections. The former is definitely needed, since usually any document has a subtitle, but it's great to see that PML also supports chapter/section subtitles because although they are not used as often they are good practice in certain types of documents (a title and subtitle can be seen as a subject and predicate, describing the section contents).

# Doc Subtitles

In any case, it's currently possible to use a `[subtitle` node with the document title:

```pml
[doc [title DOC TITLE]
     [subtitle DOC SUBTITLE]

```

which will produce the following HTML:

```html
<h1 class="pml-doc-title">DOC TITLE</h1>
<h1 class="pml-doc-title">DOC SUBTITLE</h1>
```
The problem here is double H1 headers, one after the other.

According to [§4.3.11]:

> Each heading following another heading _lead_ in the outline must have a heading level that is less than, equal to, or 1 greater than lead's heading level.

I'm not sure this disqualifies having two consecutive H1 headings as being invalid, but the point here is that semantically a subtitle should be of an inferior order to the title. Also, from a mere HTML tag point of view, it doesn't make much sense to have two separate H1 tags here, a simple `<br>` separator would have been just as fine.

But my impression, from reading the HTML5 specs, is that subtitles (in general) should be handled via the [`hgroup`][§4.3.7] element instead:

```html
<hgroup>
    <h1>Dr. Strangelove</h1>
    <p>Or: How I Learned to Stop Worrying and Love the Bomb</p>
</hgroup>
```

From [§4.3.11]  (bold added):

> The `hgroup` element represents a heading and related content. The element may be used to group an h1–h6 element with one or more p elements containing content representing a **subheading**, alternative title, or tagline.

Other markup tools don't seem to adopt this approach though.

Pandoc, when publishing books as chuncked HTML resorts to classes to handle subtitles:

```html
<header id="title-block-header">
    <h1 class="title">DOC TITLE</h1>
    <p class="subtitle">DOC SUBTITLE</p>
    <p class="author">John Doe</p>
</header>
```

I'm not sure why this is so, but probably it was in order to preserve functionality of the custom CSS created by the community. In any case, pandoc allows customizing HTML5 templates, so the above could be overridden.

Asciidoctor simply joins the document subtitle with the title, separating them with a default `:` delimiter (which can be customized). Many users complained about this, since in other output formats the document title and subtitle are styled differently (usually the latter is smaller than the former).

Either way, it might be worth investigating what the proper semantics are for handling document- and chapter-subtitles, since PML should adopt the best tagging convention in the HTML output. This is something which is best not to postpone, otherwise a later change in the HTML structure would break any existing custom CSS created by the community — something which would backlash on PML, regardless of whether these stylesheets were created for personal use or shared with the FOSS community.

I'm not an expert on HTML5, but I can clearly see that the new HTML incarnation tries to address fine-grain details of documents publishing, especially in view of how HTML/XML today are also used in e-book publishing (e.g. ePub, ePub3, and other HTML based formats). Although some of the new HTML5 tags/semantics are still not widely adopted by many popular tools, there's no reason why PML shouldn't look into them, and adopt them if they are beneficial in terms of rising the quality of the generated documents.

# Chapter Subtitles

As for chapter subtitles, the following PML source:

```pml
[ch (id=heading2sub) [title Subtitled Heading Level 2]
                    [subtitle I'm The Subtitle]
```

is rendered as:

```html
<section id="heading2sub" class="pml-chapter">
    <h2 class="pml-chapter-title">Subti</h2>
    <h2 class="pml-chapter-subtitle">I'm The Subtitle</h2>
```

again, we can see a doubling of the current heading level, just as with the doc title and subtitle. My guess is that the above considerations also apply here.

I'd love to hear from @pml-lang what the original design goals were when subtitles were implemented into PMLC, and also feedback from other users who might shed some light into the best semantics for handling titles and subtitles in various types of documents.

-------------------------------------------------------------------------------

> **NOTE** — @pml-lang please create add the `HTML templates` label to this Discussion, to simplify discovery of related topics in the future, and prevent relevant information being lost amidst the clutter.

[§4.3.11]: https://html.spec.whatwg.org/multipage/sections.html#heading-level "HTML Specs §4.3.11 Headings and outlines"
[§4.3.7]: https://html.spec.whatwg.org/multipage/sections.html#the-hgroup-element "HTML Specs §4.3.7 The hgroup element"

**Comment by pml-lang** [comment]
> From the documentation it's not clear whether a PML document supports a subtitle or not

I've locally updated the documentation (description of node `subtitle`) to make this clear.

> Each heading following another heading lead in the outline must have a heading level that is less than, *equal to*, or 1 greater than lead's heading level.
> I'm not sure this disqualifies having two consecutive H1 headings as being invalid

IMO this is allowed, because of the *"equal to"* in the specs.

> from a mere HTML tag point of view, it doesn't make much sense to have two separate H1 tags here, a simple <br> separator would have been just as fine.

That would preclude users from using different styles for titles/subtitles, and would have other diasadvantages.

> from reading the HTML5 specs, is that subtitles (in general) should be handled via the [hgroup](https://html.spec.whatwg.org/multipage/sections.html#the-hgroup-element) element instead

Thanks for mentioning this. PML should always comply with the HTML specs. Hence, we should indeed use the `hgroup` tag to wrap title/subtitle sequences. That will also allow users to provide customized CSS for title/subtitle pairs.

I suggest to do this:

For the _document_ title/subtitle, generate the following code:
```
<hgroup class="pml-doc-titles-wrapper">
    <h1 class="pml-doc-title">DOC TITLE</h1>
    <h1 class="pml-doc-subtitle">DOC SUBTITLE</h1>
</hgroup>
```

For the subtitle we could also use a `p` tag (e.g. `<p class="pml-doc-subtitle">DOC SUBTITLE</p>`), as shown in your example. But IMO a `h1` tag would be semantically more correct, because a subtitle is a 'header', not a 'paragraph'. However I'm not sure if this complies with the HTML specs.

For a _chapter_ title/subtitle, the new code would be:
```
<hgroup class="pml-chapter-titles-wrapper">
    <h2 class="pml-chapter-title">DOC TITLE</h2>
    <h2 class="pml-chapter-subtitle">DOC SUBTITLE</h2>
</hgroup>
```

PML allows to define several subtitles, which are all treated the same way. Hence, a title with two subtitles would be rendered as follows:
```
<hgroup class="pml-chapter-titles-wrapper">
    <h2 class="pml-chapter-title">DOC TITLE</h2>
    <h2 class="pml-chapter-subtitle">DOC SUBTITLE</h2>
    <h2 class="pml-chapter-subtitle">DOC SUBSUBTITLE</h2>
</hgroup>
```

Note that a `hgroup` wrapper would only be used if a title is followed by one or more subtitles.

> I'd love to hear from @pml-lang what the original design goals were when subtitles were implemented into PMLC

Subtitles were added after a user [requested](https://github.com/pml-lang/pml-companion/discussions/50) them.

> please create add the HTML templates label to this Discussion

Done.

**Reply by tajmone** [reply]
> IMO this is allowed, because of the "equal to" in the specs.

My interpretation is that it's given for granted that there would be contents between the two headings — i.e. not two adjacent headings of the same level.

From a pure editorial point of view, this:

```hmtl
<h1>Chapter</h1>
<h2>Section</h2>
```

is viewed as a chapter and a heading, i.e. they define a hierarchical structure.

but this:

```hmtl
<h1>AAAA</h1>
<h1>BBBB</h1>
```
means that Chapter "AAAA" is an empty chapter, followed by Chapter "BBBB".

I believe that the example from the specs is correct:

```html
<hgroup>
    <h1>Dr. Strangelove</h1>
    <p>Or: How I Learned to Stop Worrying and Love the Bomb</p>
</hgroup>
```

Even though it's marked as a paragraph, the text "Or: How..." is effectively a subtitle, as the given example clarifies (without margin of doubt, IMO, since that's definitely a subtitle).

> For the subtitle we could also use a `p` tag (e.g. `<p class="pml-doc-subtitle">DOC SUBTITLE</p>`), as shown in your example. But IMO a `h1` tag would be semantically more correct, because a subtitle is a 'header', not a 'paragraph'. However I'm not sure if this complies with the HTML specs.

A subtitle is not a heading, in fact it shouldn't be shown in the TOC. It's just text that better qualifies the title, like a predicate. This is true for the main title and any other heading in the contents, for when a book is catalogued only the title is indexed, and the TOC is meant to show only title headings, not their subtitles.

I'm afraid that this is just one of those aspect of proper editing rules that got corrupted by cheap tools and standards in the course of time, which replaced best practices with cheap solutions, until we reached a point when most self-improvised editors no longer distinguish between the two, and new tools are built according to common practices, instead of good practices.

It's not different from the trend to use hyphens as range separators, instead of en-dashes (`–`) or treating en- and em-dashes (`–` vs `—`) as if they were the same. Writing "1-3" instead of "1–3" is a serious typographic and editing error, and if found in a master thesis it would count as a penalty (if not to rejection of the thesis itself).

Or like the AsciiDoc rule that em-dashes must be surrounded by spaces; where does this rule come from? In some editorial context is a mistake, in other it might depend on the country and its traditions.

> PML allows to define several subtitles, which are all treated the same way. Hence, a title with two subtitles would be rendered as follows:

But why? There _can't_ be more than one subtitle for any given title, and PMLC should throw an error and abort if it encounters this since it can only be an unintentional mistake.

> Subtitles were added after a user requested them.

From the example given at #50, it's clear that what is being referred to as "subtitle 2" is actually a "preamble" (as in AsciiDoc). Different countries and publishing domains have different naming conventions and traditions regarding how Frontispieces are organized and what can go in them:

- [Wikipedia » Book design]
- [Wikipedia » Book frontispiece]
- [Wikipedia Category » Book terminology]

In any case, there's only one subtitle per title, no exceptions to the rule.

These technicalities in the implementation — which affect PML compliance to publishing standards — are the reasons why I've been insisting so much on creating official docs that present the PML vision, goals and roadmap in terms of which features and editorial standards PML will ultimately embrace.



[Wikipedia » Book design]: https://en.wikipedia.org/wiki/Book_design
[Wikipedia » Book frontispiece]: https://en.wikipedia.org/wiki/Book_frontispiece
[Wikipedia Category » Book terminology]: https://en.wikipedia.org/wiki/Category:Book_terminology

**Reply by pml-lang** [reply]
Some points to consider:

- Unfortunately there is no standard, native support for subtitles in HTML.

- The HTML spec. defines what's allowed inside a `hgroup` node. The followig peculiar [rule](https://html.spec.whatwg.org/multipage/sections.html#the-hgroup-element) in section _Content model_ for `hgroup` states:
    
	"Zero or more p elements, followed by one h1, h2, h3, h4, h5, or h6 element, followed by zero or more p elements, optionally intermixed with script-supporting elements."

	Therefore the following solution which I suggested in my previous message is invalid:

	```
	<hgroup class="pml-chapter-titles-wrapper">
		<h2 class="pml-chapter-title">CHAPTER TITLE</h2>
		<h2 class="pml-chapter-subtitle">CHAPTER SUBTITLE</h2>
	</hgroup>
	```

	The second `h2` is not allowed.

- This solution is suggested:

	```
	<hgroup>
		<h1>Dr. Strangelove</h1>
		<p>Or: How I Learned to Stop Worrying and Love the Bomb</p>
	</hgroup>
	```

	IMO this is not a good solution for two reasons:
	- A subtitle is definitely not a paragraph.
	- It would probably be challenging to correctly style the `p` node for the document and all chapter levels (e.g. font size for subtitle = 80% of font size for title).

A possible alternative solution would be this:

```
<h2 class="pml-chapter-title">CHAPTER TITLE<br/><span class="pml-chapter-subtitle">CHAPTER SUBTITLE</span></h2>
```

Note that the `span` is necessary to allow for individual styling of the subtitle.
This isn't an ideal solution neither.
Suggestions for better alternatives are welcome.

> There can't be more than one subtitle for any given title, and PMLC should throw an error and abort if it encounters this since it can only be an unintentional mistake.

Could you provide an authoritative link to this rule to point users to, if they ask for two (or more) subtitles? Or is this just an unwritten rule?
For example user coemgenuslcp [asked](https://github.com/pml-lang/pml-companion/discussions/50) for two subtitles.

> the TOC is meant to show only title headings, not their subtitles.

PMLC doesn't include subtitles in the TOC, as documented in chapter [Subtitle](https://pml-lang.dev/docs/reference_manual/index.html#node_subtitle): "The subtitle is not displayed in the table of contents."

Useful links:
- [So HGROUP Is Back In HTML 5, And Dumb As Ever!](https://medium.com/codex/so-hgroup-is-back-in-html-5-and-dumb-as-ever-c81e00f6320d)
- [The HGROUP element](https://html.spec.whatwg.org/multipage/sections.html#the-hgroup-element)
- A search for 'html chapter subtitles' reveals that there is a lot of confusion regarding subtitles in HTML


Note 1:
Section "4.12.1 Subheadings, subtitles, alternative titles and taglines" which was a sub-section of section [Common idioms without dedicated elements](https://www.w3.org/TR/2014/REC-html5-20141028/common-idioms.html#common-idioms) in an _older_ HTML5 spec. doesn't exist anymore in the [latest](https://html.spec.whatwg.org/#common-idioms) spec.


Note 2:
I used the [Nu Html Checker](https://validator.w3.org/nu/) to check my wrong code that uses a second `h2` within a `hgroup`, and it actually reported the error.
Then I used the same tool to check the HTML code genertaed for a local PML test document (uses all PML nodes), as well as for the [PML User Manual](https://pml-lang.dev/docs/user_manual/index.html). There were only a handful of very minor errors reported.
I don't know how well this tool checks _all_ HTML rules. There are other similar tools [available](https://www.softwaretestinghelp.com/html-validator/), and it would be interesting to try them out too.

**Reply by tajmone** [reply]
Unfortunately I agree with the concerns expressed by J. Knight in _[So &lt;HGROUP&gt; Is Back In HTML 5]_ regarding HTML5 being crap — this being one of the reasons I stopped working in website design when it came out. But having said this, HTML based formats remain much desired targets for tools like PML, because of their practical and widespread use for documentation as well as ebook formats like ePub (BTW, things get even uglier with ePub3).

> > There can't be more than one subtitle for any given title, and PMLC should throw an error and abort if it encounters this since it can only be an unintentional mistake.
>
> Could you provide an authoritative link to this rule to point users to, if they ask for two (or more) subtitles? Or is this just an unwritten rule?

A good baseline reference when it comes to documents styling would be the _[Chicago Manual of Style]_ (aka _CMOS_, _CMS_ or _TCM_). Although it's a reference for American English, and doesn't cover standards of specific domains like medicine, law, etc. (for which there are other standards), it's a good reference for general writing.

I'm not quite sure how you'd like me to prove the non-existence of multiple subtitles — i.e. other than the fact that we don't see them used in publications. You might have to do with the fact that when the topic is presented in style guides the singular is employed when mentioning a title and its subtitle. Most likely, no one felt the need to clarify this, since there are no expectations for it.

The term sub-title should be indicative of the fact that the element is subordinate to the title, and that any further elements would be subordinate to the subtitle instead (i.e. sub-subtitles), if we look at the hierarchy of these elements.

Trying to defend the HTML5 standard and the choices behind its "semantics" is definitely not my battle, I'm more focused on struggling to find ways to produce decent documentation with what the standard has to offer — therefore, my considerations are more _practical_ than _technical_ in nature.

I was also clear in my initial comment that I'm sharing _impressions_ about the possible uses of `<HGROUP>`, rather that facts and truths (and are perplexed as anybody else about its documentation and examples).

> IMO this is not a good solution for two reasons:
> 
> - A subtitle is definitely not a paragraph.

neither are list entries which don't qualify as full sentences, yet most tools (PML included) default to mark them as `<P>` — usually for practicality, since it helps cascading down styles in a consistent manner and less CSS definitions. There's no `<subtitle>` HTML tag, only `<title>` (which says at lot about the concerns this standard devotes to publications), so we **_have_** to find a workaround.

Anyhow, according to the _[So &lt;HGROUP&gt; Is Back In HTML 5]_ article, it seems that the intended us of `GROUP` was to then define subtitles as paragraphs, or at least this is the prevalent interpretation (and the vagueness of the documentation is most likely intention, since those who draft it didn't really know what they were doing).

It's worth pointing out here what J.Knight wrote in that article:

> What makes a two word label and a date a complete sentence, much less "one or more complete sentences encompassing a single thought"… aka what a `<p>`aragraph tag is **FOR**? Aka what a paragraph **IS**?

The implicit understand here is that for publishers the term "paragraph" refers to a full sentence which delivers meaning on its own (which is not true of a list element, for example). The title needs to be self-standing (i.e. to be meaningful even if the subtitle is left out), and the title + subtitle should form a meaningful sentence. But having said that, I see the subtitle as a qualifier of the title, so it doesn't necessary need to be a sentence on its own (although it might, and often does), and while I understand Knight's argument against marking it as a `<p>` since it doesn't qualify as a true paragraph, I think it would be a reasonable compromise to do so, in virtue of the fact that the enclosing `HGROUP` tag re-contextualized what a paragraph is here — and wouldn't be too picky about HTML5 semantics when I have to preserve the quality of the book, its styles and layout, since I've already accepted that there needs to be compromises.

Therefore, I still think that placing the subtitle in a `<P>` with a dedicated class like `<P class='subtitle'>` would provide a better solution both in terms of preserving the book structure and managing its heading styles consistently (and, YES, it does require extra CSS work, but I doubt there's a shortcut to this).

Is it a good solution? Probably not, since it only works as expected through visual styling, and might be confusing on a screen reader for the visually impaired. But then, the reality of alternative HTML representations is already permeated with violations of guidelines and rules, in order to favor usability instead. So, it's a decent workaround IMO.

When I look at the HTML source of that example, it's fairly intuitive to me that context of the paragraph being isolated within an HGROUP and immediately following a heading title implies that what I'm seeing is a subtitle (based on my knowledge that there is no `<subtitle>` tag in HTML).

The reality PML has to face is that **confusion is sovereign** when it comes to this topic of proper HTML tag choices for good publications — and the articles you linked confirm this trend.

On the one hand, there's the HTML5 standard, which is designed to encompass a very broad spectrum of applications, not just documentation, articles and books — websites, blogs, GUIs, slideshows, CVs, ... you name it — and on the other hand you have publishers, which have their well established standards and traditions. The goals of the HTML standard and publishers don't always overlap, and neither does their terminology, which often differs or carries different meanings.

For PML (as any other lightweight markup) the challenge is to find a good compromise that works well for its target audience — which, I'll never be tired to remind that it encompasses multiple standards, depending on the target domain. Which is also why I keep insisting on the need for a PML vision document, which should reassure its potential userbase on this topic. It also seems reasonable to me that the PML standard should adopt an existing style guide as its reference and blueprint, rather than relying on users requests, what you see on the web, or what other tools are doing.

We can't solve the confusion that governs HTML5, but we can counterbalance it with a solid publishing standard. This "subtitle" debate strongly exemplifies this dilemma, because even though HTML doesn't have a `<subtitle>` tag we all know that subtitles are needed in documentation and books; so, being unable to look at HTML for answers, we have to turn to style guides to find out how the "subtitle" beast is to be handled and tamed. 

The HTML is now exactly 30 years old, and my guess is that if its designers haven't felt the need for a dedicated `<subtitle>` tag in all those years the chances that they'll wake to its need in the realm of documents publications are very slim.

I'm personally not entirely happy about how AsciiDoc and pandoc approach HTML tags, because in daily use I often struggle with their limitations. Interesting to note, Asciidoctor supports subtitles in most output formats except HTML, where it simple concatenates the document title and its subtitle into a single element. I wonder if this choice was ultimately a "non choice" in light of the confusion and many practical problems being discussed here.

Even if the HTML5 standard is a pile of crap, we have no other choice than deciding which solution works best. This is nothing new in the programming field, where many (if not most) standards are either outright crap, broken, shallow or inconsistent (starting with Unicode). 

But if our quest to define good HTML tagging practices is going to be decided solely by confronting personal opinions and needs, or citing examples from third party tools, instead of adopting a well established style standard which is time tested and accepted by the publishing industry ... then the road ahead of us is a very long, tortuous and tiresome one.

-------------------------------------------------------------------------------

- [Chicago Manual of Style]
- [Chicago Manual of Style Online]



[Chicago Manual of Style Online]: https://www.chicagomanualofstyle.org/home.html "Chicago Manual of Style Online"
[Chicago Manual of Style]: https://en.wikipedia.org/wiki/The_Chicago_Manual_of_Style "Wikipedia » Chicago Manual of Style"
[So &lt;HGROUP&gt; Is Back In HTML 5]: https://medium.com/codex/so-hgroup-is-back-in-html-5-and-dumb-as-ever-c81e00f6320d

**Reply by pml-lang** [reply]
> the challenge is to find a good compromise that works well for its target audience

Yes! And make it easy for people to change the PML defaults if needed.
